### PR TITLE
feat: Add setIsArchiveDevice() method

### LINKED ISSUE #680: Add a device setting to configure what to sync
We need to persist a users's choice about what they want to sync (current options are "previews only" and "everything"). This is a device-based setting, which might be scoped to project (e.g. support different settings for each project). It shouldn't be written to the project settings of the database, which syncs with other devices.

I think this makes sense to implement on the frontend, and add an API method to change this setting.

Choices to make:

1. Default sync config for a new project instance
    a. All new project instances start with "previews only" until configured otherwise.
    b. All new project instances start with "everything" until configured otherwise.
    c. We add a new option to `api.getProject` for configuring what to sync
    d. Combination of above (option + default)
2. Method name & signature
    a. `project.configureSync('previews' | 'everything')`
    b. `project.setArchiver(true | false)`
    c. `project.setBlobDownloadFilter({ variants: Array<'preview' | 'thumbnail' | 'original'> })`
    d. Other ideas welcome!

### PR DESCRIPTION
**Pull request by gmaclennan**
Does not change sync at this time, but does persist setting.

Fixes #680

### CODE REVIEW

**Comment by gmaclennan** [comment]
Currently this throws if trying to call `manager.setIsArchiveDevice()` before first calling `setDeviceInfo()`. Not sure what the alternative is?

File: src/mapeo-project.js [hunk 1]
```diff
@@ -93,6 +95,7 @@ export class MapeoProject extends TypedEmitter {
   #loadingConfig
 
   static EMPTY_PROJECT_SETTINGS = EMPTY_PROJECT_SETTINGS
+  #isArchiveDevice
```
Comment (EvanHahn) [inline src/mapeo-project.js:98]: Nit: could you move this a few lines up, with the rest of the members?

File: test-e2e/manager-archive-device.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    async (manager2) => {
+      return manager2.getIsArchiveDevice()
+    }
+  )
+
+  assert.equal(isArchiveDevice1, false)
+  assert.equal(isArchiveDevice2, isArchiveDevice1)
+})
+
+test('migration of localDeviceInfo table', async (t) => {
```
Comment (EvanHahn) [inline test-e2e/manager-archive-device.js:84]: I love this. Nit: I think it makes more sense to test this inside a separate migrations file.

File: test-e2e/manager-archive-device.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+test('Set & Get isArchiveDevice', async () => {
+  const manager = new MapeoManager({
+    rootKey: KeyManager.generateRootKey(),
+    projectMigrationsFolder,
+    clientMigrationsFolder,
+    dbFolder: ':memory:',
+    coreStorage: () => new RAM(),
+    fastify: Fastify(),
+  })
```
Comment (EvanHahn) [inline test-e2e/manager-archive-device.js:27]: Why not use the existing test helpers for creating a manager?

**Comment by gmaclennan** [comment]
I decided to remove the requirement that `setDeviceInfo()` must be called before calling `setIsArchiveDevice()` by migrating the `deviceInfo` column to be nullable, which required manually writing the migration because SQLite does not support `ALTER COLUMN` statements. I added a test for the migration, which checks that any device info set before migration is still readable after migration.

File: test-e2e/manager-archive-device.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    project[kIsArchiveDevice],
+    true,
+    'Project isArchiveDevice is true initially'
+  )
+
+  assert.strictEqual(
+    manager.getIsArchiveDevice(),
+    true,
+    'isArchiveDevice is true initially'
+  )
```
Comment (EvanHahn) [inline test-e2e/manager-archive-device.js:42]: Nit: could you move this above the project creation to ensure that it's set immediately?

File: test-e2e/manager-archive-device.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  const projectMigrationsFolderPreMigration = fileURLToPath(
+    new URL('../drizzle/project', comapeoCorePreMigrationUrl)
+  )
+
+  const dbFolder = temporaryDirectory()
+  const rootKey = KeyManager.generateRootKey()
+  t.after(() => {
+    fs.rmSync(dbFolder, { recursive: true })
+  })
```
Comment (EvanHahn) [inline test-e2e/manager-archive-device.js:100]: Two nits:

1. could we remove this asynchronously (also requires updating the `fs` import to use `node:fs/promises`)?

```suggestion
  t.after(() => fs.rm(dbFolder, { recursive: true }))
```

2. Could we move this `t.after` call to right after we create `dbFolder`?

File: src/mapeo-project.js [hunk 6]
```diff
@@ -625,6 +631,15 @@ export class MapeoProject extends TypedEmitter {
     return deviceInfo.update(existingDoc.versionId, doc)
   }
 
+  /** @param {boolean} isArchiveDevice */
+  async [kSetIsArchiveDevice](isArchiveDevice) {
+    this.#isArchiveDevice = isArchiveDevice
+  }
+
+  get [kIsArchiveDevice]() {
```
Comment (EvanHahn) [inline src/mapeo-project.js:639]: Nit: I know the return type is trivial, but I'd love if we could add add a return type:

```suggestion
  /** @returns {boolean} */
  get [kIsArchiveDevice]() {
```

File: src/mapeo-project.js [hunk 7]
```diff
@@ -108,6 +111,7 @@ export class MapeoProject extends TypedEmitter {
    * @param {(mediaType: 'blobs' | 'icons') => Promise<string>} opts.getMediaBaseUrl
    * @param {import('./local-peers.js').LocalPeers} opts.localPeers
    * @param {Logger} [opts.logger]
+   * @param {boolean} [opts.isArchiveDevice] Whether this device is an archive device
```
Comment (EvanHahn) [inline src/mapeo-project.js:114]: Is there ever a time we'd instantiate a project without this option? I think not, so maybe we can make this required?

**Review by EvanHahn — CHANGES_REQUESTED** [review]
LGTM other than a few nitpicks and small questions.

File: src/mapeo-project.js (hunk 7 above)
Comment (gmaclennan) [reply src/mapeo-project.js]: Addressed in 2dc69a6a0351cb1db0ff65b813e1b0efdccaccae

File: test-e2e/manager-archive-device.js (hunk 3 above)
Comment (gmaclennan) [reply test-e2e/manager-archive-device.js]: Addressed in 2dc69a6a0351cb1db0ff65b813e1b0efdccaccae

**Review by EvanHahn — APPROVED** [review]

File: test-e2e/manager-archive-device.js (hunk 2 above)
Comment (gmaclennan) [reply test-e2e/manager-archive-device.js]: Addressed in 2dc69a6a0351cb1db0ff65b813e1b0efdccaccae

File: src/mapeo-project.js (hunk 1 above)
Comment (gmaclennan) [reply src/mapeo-project.js]: Addressed in 2dc69a6a0351cb1db0ff65b813e1b0efdccaccae

**Pull request merged**